src/coreclr/jit/lclmasks.cpp

```diff
@@ -0,0 +1,480 @@
+// in the convert and the weighting of the block it exists in. The second value assumes the
+// local var has been switched to store as a mask and performs the same count. The switch
+// will count removes every existing convert and add a convert where there isn't currently
+// a convert.
+//
+// Once every definition and use has been parsed, the parsing runs again. At each step,
+// if the weighting for switching that var is lower than the current weighting then switch
+// to store as mask and add/remove conversions as required.
+//
+// Limitations:
+//
+// Local variables that are defined then immediately used just once may not be saved to a
+// store. Here a convert to to vector will be used by a convert to mask. These instances will
+// be caught in the lowering phase.
+//
+// This weighting does not account for re-definition. A variable may first be created as a
+// mask used as such, then later in the method defined as a vector and used as such from
+// then on. This can be worked around at the user level by encouraging users not to reuse
+// variable names.
+//
+// Returns:
+//    Suitable phase status
+//
+PhaseStatus Compiler::fgOptimizeLclMasks()
+{
+#if defined(TARGET_ARM64)
+
+    if (opts.OptimizationDisabled())
+    {
+        JITDUMP("Skipping. Optimizations Disabled\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    if (!compMaskConvertUsed)
+    {
+        JITDUMP("Skipping. There are no converts of locals \n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    LclMasksWeightTable weightsTable = LclMasksWeightTable(getAllocator());
+
+    // Find every local and add them to weightsTable.
+    bool foundConversion = false;
+    JITDUMP("\n");
+    for (BasicBlock* block : Blocks())
+    {
+        for (Statement* const stmt : block->Statements())
+        {
+            GenTreeLclVarCommon* firstLcl = *stmt->LocalsTreeList().begin();
+            if (firstLcl != nullptr)
+            {
+                LclMasksCheckVisitor ev(this, block->getBBWeight(this), &weightsTable);
+                GenTree*             root = stmt->GetRootNode();
+                ev.WalkTree(&root, nullptr);
+                foundConversion |= ev.foundConversions;
+            }
+        }
+    }
+
+    if (!foundConversion)
+    {
+        JITDUMP("Done. No conversions of locals found.\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    // For each Local, potentially add/remove a conversion.
+    JITDUMP("\n");
+    for (BasicBlock* block : Blocks())
+    {
+        for (Statement* const stmt : block->Statements())
+        {
+            GenTreeLclVarCommon* firstLcl = *stmt->LocalsTreeList().begin();
+            if (firstLcl != nullptr)
+            {
+                LclMasksUpdateVisitor ev(this, stmt, &weightsTable);
+                GenTree*              root = stmt->GetRootNode();
+                ev.WalkTree(&root, nullptr);
```

Comment:
I think you should do the `fgSequenceLocals` from here, in the case where any changes were made in the statement. No reason to do it potentially multiple times from inside the visitor.